GamesDoneQuick/donation-tracker
# add Admin diagnostic page [#174782375]

# Contributing to the Donation Tracker

First of all, thank you for taking the time to contribute!

Please fill out the template below and check the following boxes:

- [X] I've added tests or modified existing tests for the change.
- [X] I've humanly end-to-end tested the change by running an instance of the tracker.

### Issue from Pivotal Tracker

https://www.pivotaltracker.com/story/show/174782375

### Description of the Change

Rather than expose public URLs that only work if debug is on, this combines them into a Diagnostic page which also adds some additional checks/tests/button to verify that other things are working right. Basically if I think they're tricky to set up for any reason, they end up here. You can only use it if you're a superuser.

Also, I moved the tracker admin templates into a subfolder to avoid namespace clashes.

Additionally, I discovered a weird quirk of the way Celery and Post Office work that meant trying to run the tests on my machine always hung forever(?) attempting to queue up a task on a RabbitMQ instance that didn't exist. The test runner now sets up Celery to always execute tasks immediately.

### Verification Process

There are currently five bits to this:

- Open the ping websocket (used to be located at `/tracker/websocket_test`
- Open the celery websocket and queue a task (used to be located at `/tracker/celery_test`
- Send a quick and dirty test email (really just uses Post Office, but it's nice to have a one-click for this)
- Shows if Django thinks the request is secure (there's a lot going on here, but I was able to get it to show `True` by going through ngrok)
- Shows if Django is able to write/read/delete a test file with whatever the current file storage setup is
  - worked with the default file storage
  - set up various stages of "S3 is only partially set up" and got the expected errors, and eventually got a success
